design/baremetal-operator/bmh-v1beta1.md

```diff
@@ -0,0 +1,140 @@
+itself and in Ironic.
+
+For example, Ironic is planning a complete removal of MD5 support for
+checksums.
+
+### Goals
+
+- Clean up issues in the BareMetalHost API that have accumulated during
+  the usage of `v1alpha1`.
+
+- Make the API version reflect the actual support status of the API (see
+  [alternatives](#Alternatives) for details).
+
+### Non-Goals
+
+- Stabilize the BareMetalHost API completely (as `v1`).
+
+## Proposal
+
+Copy the current API types definition into a new package `v1beta1` with some
+modifications.
+
+## Design Details
+
+The following addition should be done *before* the version is created:
+
+- Add a `Spec.Architecture` field to override architecture in the profile.
+  If not provided, BMO will use its own architecture as the default.
+
+The following changes will be done to the new version:
+
+- Remove MD5 support from `Image.ChecksumType`. When no `ChecksumType` is
+  provided, do not populate the `os_image_hash_algo` field and let Ironic
+  guess the algorithm.
+
+  The same change is applied to CAPM3.
+
+- In `RootDeviceHints`:
+
+  Rename `DeviceName` to `DevicePath` to better reflect what it actually is
+  (especially after introducing support for `/dev/disk/by-path`).
+
+  Rename `Model` and `Vendor` to `ModelContains` and `VendorContains` to
+  reflect the fact that these fields use the `<in>` operator.
+
+- Remove `HardwareProfile` and don't do any matching.
+
+- Remove `HardwareDetails` from the `Status` in favour of the already
+  introduced `HardwareData` CR.
+
+- Rename `online` to `poweredOn`.
+
+  The word *online* is ambiguous here: it may be understood to refer to
+  provisioning status. Furthermore, the corresponding field in the `Status`
+  is already called `poweredOn`.
+
+### Implementation Details/Notes/Constraints
+
+Both `v1alpha1` and the proposed `v1beta1` resources will be possible
+to handle with the same controller code since the differences are mostly
+cosmetic.
+
+### Risks and Mitigations
+
+None
+
+### Work Items
+
+- Add `Architecture`
+- Refactor the BMO code to avoid referring directly to the current version.
+- Copy type definitions and modify them.
+- Create a conversion webhook to change between versions.
+- Communicate the change to the community.
+
+### Dependencies
+
+None
+
+### Test Plan
+
+We'll need to test both API versions. Since we already have tests for different
+versions of CAPM3, we may tie them to different versions of BMH.
+
+### Upgrade / Downgrade Strategy
+
+The conversion webhook will take care of in-flight CR upgrade:
+
+- Deleting `HardwareProfile`, populating `Architecture` and `RootDeviceHints`
+  if needed.
+
+- Removing `ChecksumType` if its value is `MD5`. For a while, Ironic will
+  still accept MD5 checksums.
+
+### Version Skew Strategy
+
+BMO and CAPM3 version skew may affect the `ChecksumType` handling. Even
+with older CAPM3, users will not be able to use `MD5` as the value, but
+the other values will be available.
+
```

Comment:
We have discussed backporting v1beta1 BMO API support to older CAPM3 release branches and we are on the opinion (ATM), that only the latest release branch would get a backport to support v1beta1 of bmo, or there would be a completely new minor release of CAPM3 that could use the new BMO API but old CAPM3's would only work with old BMO and v1alpha1 API. I guess this info can be incorporated in the proposal and this would solve the version skew issues.

I guess if the MD5  related api changes are only done to the v1beta1 BMO api and a user using an old BMO version with an old CAPM3 version with an old Ironic everything would be fine.